### DISCUSSION TITLE
declared_attr and type hints

### DISCUSSION THREAD

**Discussion by MatthewNewland** [sqlalchemy/sqlalchemy#10008] [Usage Questions] [answered]
Hello,

This is related to the [Litestar](https://github.com/litestar-org/litestar) project which ships a few preconfigured SQLAlchemy base classes.

For their integer primary key classes, they use a mixin, which simplified looks as follows:

```python
BigIntIdentity = BigInteger().with_variant(Integer, "sqlite")

class BigIntPrimaryKey:
    """BigInt Primary Key Field Mixin."""

    @declared_attr
    def id(cls) -> Mapped[int]:
        """BigInt Primary key column."""
        return mapped_column(
            BigIntIdentity,
            Sequence(f"{cls.__tablename__}_id_seq", optional=False),
            primary_key=True,
        )
```  
(See https://github.com/litestar-org/litestar/tree/main/litestar/contrib/sqlalchemy if you want more context for the code, or feel free to ask.)

Now, they also use `typing.get_type_hints` on 2.0 style models to help with data parsing (see the `dto.py` in the folder above), but the descriptors that are defined through `declared_attr` do not seem to return anything through `get_type_hints`.

My question is: What is the preferred way to get the typing for a descriptor declared this way?

Please let me know if any of this doesn't make sense!

**Comment by zzzeek** [comment, marked as answer]
get_type_hints() is not sufficient to find out about every annotation on a class.    These annotations are also on the attributes themselves for things like methods, so any parsing would have to take that into account:

```py
(Pdb) BigIntPrimaryKey.__annotations__
{}
(Pdb) BigIntPrimaryKey.__dict__['id'].fget.__annotations__
{'return': 'Mapped[int]'}

```

Above, `@declared_attr` is a decorator, which uses the "fget" name just like [standard Python @property](https://docs.python.org/3/library/functions.html#property).  that's where the function is.     you would have to do something similar for plain `@property`:

```py
class BigIntPrimaryKey:
    @property
    def id(self) -> Mapped[int]:
        ...

```
same thing:

```py
(Pdb) BigIntPrimaryKey.__dict__['id']
<property object at 0x7fc890927920>
(Pdb) BigIntPrimaryKey.__dict__['id'].fget
<function BigIntPrimaryKey.id at 0x7fc88f91b4c0>
(Pdb) BigIntPrimaryKey.__dict__['id'].fget.__annotations__
{'return': 'Mapped[int]'}

```


runtime introspection for types is not that easy so a lot of trial-and-error has to be employed.
